File: ios/AutoFans/Core/MobileAPI.swift
import Foundation
import OSLog
import UIKit

actor MobileAPI {
    private let config: APIConfiguration
    private let auth: AuthStore
    private let encoder = JSONEncoder(), decoder = JSONDecoder()
    private static let logger = Logger(subsystem: "ro.autofans.app", category: "MobileAPI")
    private static let activitySessionKey = "autofans.activity.session"
    init(config: APIConfiguration, auth: AuthStore) { self.config = config; self.auth = auth }

    func call(_ operation: String, payload: [String: JSONValue] = [:]) async throws -> [String: JSONValue] {
        let function = Self.chatOperations.contains(operation) ? "chat-v1" : "mobile-v1"
        Self.logger.info("Preparing \(operation, privacy: .public) via \(function, privacy: .public)")
        let session: AuthSession
        do {
            session = try await auth.activeSession()
        } catch {
            Self.logger.error("\(operation, privacy: .public) could not get an active session: \(error.localizedDescription, privacy: .public)")
            throw error
        }
        var request = URLRequest(url: config.supabaseURL.appendingPath("functions/v1/\(function)"))
        request.httpMethod = "POST"
        request.timeoutInterval = 20
        request.setValue(config.anonKey, forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(session.accessToken)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try encoder.encode(["operation": JSONValue.string(operation), "payload": JSONValue.object(payload)])

        Self.logger.info("Starting \(operation, privacy: .public) via \(function, privacy: .public)")
        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse else {
                Self.logger.error("\(operation, privacy: .public) returned a non-HTTP response")
                throw APIError.invalidResponse
            }
            guard 200..<300 ~= http.statusCode else {
                let message = SupabaseRepository.message(data) ?? "Cererea a eșuat."
                Self.logger.error("\(operation, privacy: .public) failed with HTTP \(http.statusCode): \(message, privacy: .public)")
                throw APIError.server(status: http.statusCode, message: message)
            }
            let result = try decoder.decode([String: JSONValue].self, from: data)
            Self.logger.info("Finished \(operation, privacy: .public) with HTTP \(http.statusCode)")
            return result
        } catch {
            Self.logger.error("\(operation, privacy: .public) request failed: \(error.localizedDescription, privacy: .public)")
            throw error
        }
    }

    func registerPushToken(_ token: String) async throws {
        guard token.count >= 30 else { return }
        _ = try await call("register_push_token", payload: [
            "token": .string(token),
            "platform": .string("ios"),
        ])
    }

    func uploadListingImage(_ image: UIImage) async throws -> String {
        guard let data = image.jpegData(compressionQuality: 0.82), data.count <= 10 * 1024 * 1024 else { throw APIError.server(status: 400, message: "Imaginea poate avea cel mult 10 MB.") }
        let session = try await auth.activeSession(), path = "\(session.user.id)/mobile/\(UUID().uuidString).jpg"
        var request = URLRequest(url: config.supabaseURL.appendingPath("storage/v1/object/listing-images/\(path)")); request.httpMethod = "POST"; request.setValue(config.anonKey, forHTTPHeaderField: "apikey"); request.setValue("Bearer \(session.accessToken)", forHTTPHeaderField: "Authorization"); request.setValue("image/jpeg", forHTTPHeaderField: "Content-Type"); request.httpBody = data
        let (_, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, 200..<300 ~= http.statusCode else { throw APIError.server(status: (response as? HTTPURLResponse)?.statusCode ?? 0, message: "Nu am putut încărca imaginea.") }
        return path
    }

    /// Compresses and uploads a profile image to the public avatar bucket. The
    /// storage policy allows each user to write only inside their own UUID folder.
    func uploadProfileAvatar(_ image: UIImage) async throws -> String {
        let maximumDimension: CGFloat = 1_200
        let sourceSize = image.size
        let scale = min(1, maximumDimension / max(sourceSize.width, sourceSize.height))
        let targetSize = CGSize(width: max(1, floor(sourceSize.width * scale)), height: max(1, floor(sourceSize.height * scale)))
        let rendered = UIGraphicsImageRenderer(size: targetSize).image { _ in image.draw(in: CGRect(origin: .zero, size: targetSize)) }

        guard let data = rendered.jpegData(compressionQuality: 0.82), data.count <= 5 * 1024 * 1024 else {
            throw APIError.server(status: 400, message: "Fotografia de profil poate avea cel mult 5 MB.")
        }

        let session = try await auth.activeSession()
        let path = "\(session.user.id)/avatar-\(UUID().uuidString).jpg"
        var request = URLRequest(url: config.supabaseURL.appendingPath("storage/v1/object/profile-avatars/\(path)"))
        request.httpMethod = "POST"
        request.setValue(config.anonKey, forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(session.accessToken)", forHTTPHeaderField: "Authorization")
        request.setValue("image/jpeg", forHTTPHeaderField: "Content-Type")
        request.httpBody = data

        let (_, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, 200..<300 ~= http.statusCode else {
            throw APIError.server(status: (response as? HTTPURLResponse)?.statusCode ?? 0, message: "Nu am putut încărca fotografia de profil.")
        }
        return config.supabaseURL.appendingPath("storage/v1/object/public/profile-avatars/\(path)").absoluteString
    }

    func recordListingView(listingID: Int) async throws {
        try await recordPublicActivity(
            path: "rest/v1/listing_views",
            payload: [
                "listing_id": .number(Double(listingID)),
                "visitor_id": .null,
                "session_id": .string(activitySessionID()),
                "viewed_on": .string(Self.currentActivityDay()),
            ],
        )
    }

    func recordListingContact(listingID: Int, contactType: String) async throws {
        let normalizedType = Self.normalizedContactType(contactType)
        try await recordPublicActivity(
            path: "rest/v1/listing_contacts",
            payload: [
                "listing_id": .number(Double(listingID)),
                "visitor_id": .null,
                "contact_type": .string(normalizedType),
                "session_id": .string(activitySessionID()),
                "contacted_on": .string(Self.currentActivityDay()),
            ],
        )
    }

    private static let chatOperations: Set<String> = [
        "conversations",
        "messages",
        "start_conversation",
        "send_message",
        "mark_conversation_read",
    ]

    private func activitySessionID() -> String {
        if let existing = UserDefaults.standard.string(forKey: Self.activitySessionKey), existing.count >= 12 {
            return existing
        }
        let created = UUID().uuidString.lowercased()
        UserDefaults.standard.set(created, forKey: Self.activitySessionKey)
        return created
    }

    private static func currentActivityDay() -> String {
        ISO8601DateFormatter().string(from: Date()).prefix(10).description
    }

    private static func normalizedContactType(_ value: String) -> String {
        switch value {
        case "phone", "whatsapp", "viewing":
            return value
        default:
            return "message"
        }
    }

    private func recordPublicActivity(path: String, payload: [String: JSONValue]) async throws {
        var request = URLRequest(url: config.supabaseURL.appendingPath(path))
        request.httpMethod = "POST"
        request.timeoutInterval = 10
        request.setValue(config.anonKey, forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(config.anonKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("return=minimal", forHTTPHeaderField: "Prefer")
        request.httpBody = try encoder.encode(payload)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw APIError.invalidResponse }
        if 200..<300 ~= http.statusCode { return }

        if http.statusCode == 409 {
            let duplicate = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["code"] as? String
            if duplicate == "23505" { return }
        }

        throw APIError.server(status: http.statusCode, message: SupabaseRepository.message(data) ?? "Nu am putut actualiza statisticile anunțului.")
    }

}

File: android/app/src/main/java/ro/autofans/app/data/MobileApi.kt
package ro.autofans.app.data

import android.content.Context
import android.net.Uri
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.HttpUrl.Companion.toHttpUrl
import java.time.LocalDate
import java.io.IOException
import java.util.UUID

/** Authenticated companion to the public listing repository. All mutations
 * travel through the server façade so Android and a later Swift client share
 * input validation and error shapes. */
class MobileApi(
    private val config: SupabaseConfig,
    private val auth: SupabaseAuthRepository,
    private val clientSessionId: String,
    private val client: OkHttpClient = OkHttpClient(),
    private val json: Json = Json { ignoreUnknownKeys = true },
) {
    fun currentUserId(): String? = auth.session?.user?.id

    suspend fun subscribeToMessages(onEvent: (RealtimeMessageEvent) -> Unit, onError: (Throwable) -> Unit): java.io.Closeable = withContext(Dispatchers.IO) {
        val session = auth.activeSession()
        val websocketUrl = config.url.toHttpUrl().newBuilder()
            .scheme(if (config.url.startsWith("https://")) "wss" else "ws")
            .addPathSegments("realtime/v1/websocket")
            .addQueryParameter("apikey", config.anonKey)
            .addQueryParameter("vsn", "1.0.0")
            .build()
        val (listener, makeSubscription) = realtimeMessageListener(json, session.access_token, onEvent, onError)
        val socket = client.newWebSocket(
            Request.Builder().url(websocketUrl).header("apikey", config.anonKey).header("Authorization", "Bearer ${session.access_token}").build(),
            listener,
        )
        makeSubscription(socket)
    }

    suspend fun call(operation: String, payload: JsonObject = buildJsonObject {}): JsonObject = withContext(Dispatchers.IO) {
        val session = auth.activeSession()
        val body = buildJsonObject { put("operation", operation); put("payload", payload) }.toString()
        val functionName = if (operation in CHAT_OPERATIONS) "chat-v1" else "mobile-v1"
        val request = Request.Builder().url("${config.url.trimEnd('/')}/functions/v1/$functionName")
            .header("apikey", config.anonKey)
            .header("Authorization", "Bearer ${session.access_token}")
            .post(body.toRequestBody(JSON)).build()
        client.newCall(request).execute().use { response ->
            val text = response.body?.string().orEmpty()
            if (!response.isSuccessful) throw SupabaseException(response.code, mobileApiError(response.code, response.header("Content-Type"), text))
            return@withContext runCatching { json.parseToJsonElement(text) as? JsonObject }.getOrNull()
                ?: throw IOException("Serverul AutoFans a trimis un răspuns invalid. Încearcă din nou mai târziu.")
        }
    }

    /** Registers the current physical device for server-side FCM delivery.
     * The FCM token is never a credential and is always tied to the active
     * authenticated AutoFans user by the Edge Function. */
    suspend fun registerPushToken(token: String) {
        if (token.length < 30) return
        call("register_push_token", buildJsonObject {
            put("token", token)
            put("platform", "android")
        })
    }

    suspend fun uploadListingImage(context: Context, uri: Uri): String = withContext(Dispatchers.IO) {
        val session = auth.activeSession()
        val bytes = context.contentResolver.openInputStream(uri)?.use { it.readBytes() } ?: error("Nu am putut citi imaginea.")
        require(bytes.size <= 10 * 1024 * 1024) { "Imaginea poate avea cel mult 10 MB." }
        val path = "${session.user.id}/mobile/${UUID.randomUUID()}.jpg"
        val request = Request.Builder().url("${config.url.trimEnd('/')}/storage/v1/object/listing-images/$path")
            .header("apikey", config.anonKey).header("Authorization", "Bearer ${session.access_token}")
            .header("Content-Type", context.contentResolver.getType(uri) ?: "image/jpeg")
            .post(bytes.toRequestBody()).build()
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) throw SupabaseException(response.code, response.body?.string().orEmpty())
        }
        path
    }

    /** Uploads an avatar to its dedicated public bucket. The app only receives
     * a public URL after Storage has accepted an authenticated upload. */
    suspend fun uploadProfileAvatar(context: Context, uri: Uri): String = withContext(Dispatchers.IO) {
        val session = auth.activeSession()
        val contentType = context.contentResolver.getType(uri) ?: "image/jpeg"
        require(contentType in PROFILE_AVATAR_MIME_TYPES) { "Alege o imagine JPEG, PNG sau WebP." }
        val bytes = context.contentResolver.openInputStream(uri)?.use { it.readBytes() }
            ?: error("Nu am putut citi fotografia.")
        require(bytes.size <= PROFILE_AVATAR_MAX_BYTES) { "Fotografia de profil poate avea cel mult 5 MB." }
        val extension = when (contentType) {
            "image/png" -> "png"
            "image/webp" -> "webp"
            else -> "jpg"
        }
        val path = "${session.user.id}/avatar-${UUID.randomUUID()}.$extension"
        val request = Request.Builder().url("${config.url.trimEnd('/')}/storage/v1/object/profile-avatars/$path")
            .header("apikey", config.anonKey)
            .header("Authorization", "Bearer ${session.access_token}")
            .header("Content-Type", contentType)
            .post(bytes.toRequestBody()).build()
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) throw SupabaseException(response.code, response.body?.string().orEmpty())
        }
        "${config.url.trimEnd('/')}/storage/v1/object/public/profile-avatars/$path"
    }

    suspend fun signListingImageUrls(paths: List<String>): Map<String, String> = withContext(Dispatchers.IO) {
        val distinctPaths = paths.filter(String::isNotBlank).distinct()
        if (distinctPaths.isEmpty()) return@withContext emptyMap()
        val session = auth.activeSession()
        val requestJson = json.encodeToString(SignRequest(distinctPaths, 3_600))
        val request = Request.Builder()
            .url("${config.url.trimEnd('/')}/storage/v1/object/sign/listing-images")
            .header("apikey", config.anonKey)
            .header("Authorization", "Bearer ${session.access_token}")
            .post(requestJson.toRequestBody(JSON))
            .build()
        client.newCall(request).execute().use { response ->
            val text = response.body?.string().orEmpty()
            if (!response.isSuccessful) throw SupabaseException(response.code, mobileApiError(response.code, response.header("Content-Type"), text))
            json.decodeFromString<List<SignedUrlDto>>(text).mapNotNull { signed ->
                val path = signed.path ?: signed.name ?: return@mapNotNull null
                val signedPath = signed.signedUrl ?: return@mapNotNull null
                val fullUrl = if (signedPath.startsWith("http")) signedPath else "${config.url.trimEnd('/')}/storage/v1$signedPath"
                path to fullUrl
            }.toMap()
        }
    }

    suspend fun recordListingView(listingId: Long) {
        recordPublicActivity(
            table = "listing_views",
            payload = buildJsonObject {
                put("listing_id", listingId)
                put("visitor_id", kotlinx.serialization.json.JsonNull)
                put("session_id", clientSessionId)
                put("viewed_on", currentActivityDay())
            },
        )
    }

    suspend fun recordListingContact(listingId: Long, contactType: String) {
        recordPublicActivity(
            table = "listing_contacts",
            payload = buildJsonObject {
                put("listing_id", listingId)
                put("visitor_id", kotlinx.serialization.json.JsonNull)
                put("contact_type", normalizeContactType(contactType))
                put("session_id", clientSessionId)
                put("contacted_on", currentActivityDay())
            },
        )
    }

    private companion object {
        val CHAT_OPERATIONS = setOf("conversations", "messages", "start_conversation", "send_message", "mark_conversation_read")
        val JSON = "application/json; charset=utf-8".toMediaType()
        val PROFILE_AVATAR_MIME_TYPES = setOf("image/jpeg", "image/png", "image/webp")
        const val PROFILE_AVATAR_MAX_BYTES = 5 * 1024 * 1024
    }

    @Serializable
    private data class SignRequest(val paths: List<String>, @SerialName("expiresIn") val expiresIn: Int)

    @Serializable
    private data class SignedUrlDto(
        val name: String? = null,
        val path: String? = null,
        @SerialName("signedURL") val signedUrl: String? = null,
    )

    private suspend fun recordPublicActivity(table: String, payload: JsonObject) = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("${config.url.trimEnd('/')}/rest/v1/$table")
            .header("apikey", config.anonKey)
            .header("Authorization", "Bearer ${config.anonKey}")
            .header("Content-Type", "application/json")
            .header("Prefer", "return=minimal")
            .post(payload.toString().toRequestBody(JSON))
            .build()
        client.newCall(request).execute().use { response ->
            val text = response.body?.string().orEmpty()
            if (response.isSuccessful) return@withContext
            if (response.code == 409 && text.contains("\"code\":\"23505\"")) return@withContext
            throw SupabaseException(response.code, mobileApiError(response.code, response.header("Content-Type"), text))
        }
    }

    private fun currentActivityDay(): String = LocalDate.now().toString()

    private fun normalizeContactType(value: String): String = when (value) {
        "phone", "whatsapp", "viewing" -> value
        else -> "message"
    }

    private fun mobileApiError(status: Int, contentType: String?, body: String): String {
        val isHtml = contentType?.contains("text/html", ignoreCase = true) == true || body.trimStart().startsWith("<!DOCTYPE", ignoreCase = true)
        return when {
            status == 404 && isHtml -> "Serviciul mobil AutoFans nu este încă disponibil pe server. Actualizează aplicația după ce serverul a fost publicat."
            isHtml -> "Serverul AutoFans a trimis o pagină web în locul răspunsului aplicației. Încearcă din nou mai târziu."
            else -> body.ifBlank { "Cererea către AutoFans a eșuat (cod $status)." }
        }
    }
}
